llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -17577,3 +17577,13 @@ SITargetLowering::lowerIdempotentRMWIntoFencedLoad(AtomicRMWInst *AI) const {
   AI->eraseFromParent();
   return LI;
 }
+
+bool SITargetLowering::hasAndNot(SDValue Op) const {
+  // AND-NOT is only valid on uniform (SGPR) values; divergent values live in
+  // VGPRs.
+  if (Op->isDivergent())
+    return false;
+
+  EVT VT = Op.getValueType();
+  return VT == MVT::i32 || VT == MVT::i64;
```

Comment:
I'm not sure we need to check types here. How about just `return !Op->isDivergent();`? If the types are not legal they will get legalized, but that should not affect the decision of whether to form an and-not pattern.

Reply:
If it's a different type and then is legalized, there will be intermediate instructions that break the and not pattern